### PR TITLE
remove the nat synonyms feature from the compiler

### PR DESCRIPTION
**Pull request by paulbrauner-da**
Context: https://github.com/digital-asset/daml/issues/18240

Remove any usage of the feature, in code and in the integration tests.

### CODE REVIEW

**Review by remyhaemmerle-da — APPROVED** [review]

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs [hunk 1]
```diff
@@ -169,14 +169,6 @@ featureExceptions = Feature
     , featureCppFlag = Just "DAML_EXCEPTIONS"
     }
 
-featureNatSynonyms :: Feature
-featureNatSynonyms = Feature
-    { featureName = "Nat type synonyms"
-    , featureVersionReq = VersionReq \case
-          V2 -> allMinorVersions
-    , featureCppFlag = Just "DAML_NAT_SYN"
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:177]: Remy just confirmed seven minutes ago :D

**Review by akrmn — APPROVED** [review]
LGTM, thanks @paulbrauner-da!

For future readers, `data NatSyn n` is kept, it is the versioned "feature" that's getting removed

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulbrauner-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:177]: Good point, I always perceived these flags as internal. @remyhaemmerle-da what do you think?

**Comment by paulbrauner-da** [comment]
Good point, let me rephrase the title as "feature flag"

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs [hunk 2]
```diff
@@ -247,6 +247,13 @@ featureExperimental = Feature
     , featureCppFlag = Just "DAML_EXPERIMENTAL"
     }
 
+-- | CPP flags of past features that have become part of LF but that some
+-- clients might still depend on being defined.
+foreverCppFlags :: [T.Text]
+foreverCppFlags =
```
Comment (dylant-da) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:253]: Looks good :) Should we create an error in compiler/damlc/daml-prim-src/DA/Internal/NatSyn.daml if DAML_NAT_SYN is not defined, since it always should be when we build prim from now on?

Comment (paulbrauner-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Version.hs:253]: Addressed in https://github.com/digital-asset/daml/pull/18246/commits/23b03eeaea3b501bac4d87cf4f2093437935417e

File: compiler/damlc/tests/daml-test-files/LegacyCppFlags.daml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+-- Dummy script to prevent a "redundant import" error.
+dummyScript : Script ()
+dummyScript = pure ()
+
+#ifndef DAML_NAT_SYN
+natSynNotDefined : Script ()
+natSynNotDefined =
+  error "Expected DAML_NAT_SYN to be defined for backwards compatibility"
+#endif
```
Comment (akrmn) [inline compiler/damlc/tests/daml-test-files/LegacyCppFlags.daml:15]: How about

```suggestion
test : Script ()
#ifndef DAML_NAT_SYN
  error "Expected DAML_NAT_SYN to be defined for backwards compatibility"
#endif
  pure ()
```

File: compiler/damlc/tests/daml-test-files/LegacyCppFlags.daml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{-# LANGUAGE CPP #-}
+
+module LegacyCppFlags where
+
+import Daml.Script
+
+test : Script ()
+#ifndef DAML_NAT_SYN
```
Comment (akrmn) [inline compiler/damlc/tests/daml-test-files/LegacyCppFlags.daml:8]: my bad, I forgot to add `test = do`
```suggestion
test : Script ()
test = do
#ifndef DAML_NAT_SYN
```

Comment (paulbrauner-da) [reply compiler/damlc/tests/daml-test-files/LegacyCppFlags.daml]: No worries, fixed in 4453ca3
